src/OrbitGgp/Client.cpp

```diff
@@ -159,6 +165,28 @@ void Client::GetInstancesAsync(
       });
 }
 
+Future<ErrorMessageOr<QByteArray>> Client::RunGgpProcess(QStringList arguments) {
+  return thread_pool_->Schedule(
+      [this, arguments = std::move(arguments)]() -> ErrorMessageOr<QByteArray> {
+        QProcess process;
+        process.start(ggp_program_, arguments, QIODevice::ReadOnly);
+
+        if (!process.waitForFinished(timeout_.count()) && process.exitCode() == 0) {
+          return ErrorMessage(
+              absl::StrFormat("Ggp cli process timed out after %dms", timeout_.count()));
+        }
+
+        if (process.exitCode() != 0) {
+          return ErrorMessage(absl::StrFormat(
+              "Ggp cli process failed, exit code: %d, error: %s, stdout: %s, stderr: %s",
+              process.exitCode(), process.errorString().toStdString(),
+              process.readAllStandardOutput().data(), process.readAllStandardError().data()));
+        }
+
+        return process.readAll();
+      });
```

Comment:
This is not ideal because the thread pool could be blocked by a couple of long running Ggp processes. And it is also not needed. You could do something like that:

```
orbit_base::Promise<ErrorMessageOr<QByteArray>> promise{};
auto future = promise.GetFuture();
auto process = new Process{};
QObject::connect(&process, &QProcess::finished, [promise = std::move(promise), process](int exit_code, QProcess::ExitStatus exit_status) {
  promise.SetResult(...);
  process->deleteLater();
});
process->start*...);
return future;
```

This is missing the handling of the timeout via QTimer but I guess you get the idea.

(I will also look at the rest - but putting this out for now)